typescript/infra/src/funding/adapters/EVMFundingAdapter.ts

```diff
@@ -0,0 +1,415 @@
+import { EthBridger, getArbitrumNetwork } from '@arbitrum/sdk';
+import { CrossChainMessenger } from '@eth-optimism/sdk';
+import { BigNumber, Signer, ethers } from 'ethers';
+import { Logger } from 'pino';
+import { Gauge } from 'prom-client';
+import { format } from 'util';
+
+import {
+  BaseEvmAdapter,
+  ChainName,
+  HyperlaneIgp,
+  MultiProtocolProvider,
+  MultiProvider,
+} from '@hyperlane-xyz/sdk';
+import { Address } from '@hyperlane-xyz/utils';
+
+import { Contexts } from '../../../config/contexts.js';
+import L1ETHGateway from '../../../scripts/funding/utils/L1ETHGateway.json' with { type: 'json' };
+import L1MessageQueue from '../../../scripts/funding/utils/L1MessageQueue.json' with { type: 'json' };
+import L1ScrollMessenger from '../../../scripts/funding/utils/L1ScrollMessenger.json' with { type: 'json' };
+import { BaseAgentKey } from '../../agents/keys.js';
+import { DeployEnvironment } from '../../config/environment.js';
+import { FundableRole } from '../../roles.js';
+import { L1_CHAIN, L2_CHAINS } from '../helpers.js';
+import { FundingAddresses } from '../types.js';
+
+import { IFundingAdapter } from './IFundingAdapter.js';
+
+/**
+ * EVM-specific implementation of the funding adapter
+ * Extends BaseEvmAdapter to leverage EVM-specific functionality
+ */
+export class EVMFundingAdapter
+  extends BaseEvmAdapter
+  implements IFundingAdapter
+{
+  private readonly localMultiProvider: MultiProvider;
+  private readonly igp: HyperlaneIgp;
+  private readonly nativeBridges: Record<
+    string,
+    { l1ETHGateway: string; l1Messenger: string }
+  >;
+  private readonly L2Chains: ChainName[];
+  private readonly L2ToL1: Record<ChainName, ChainName>;
+
+  constructor(
+    chainName: ChainName,
+    private readonly multiProtocolProvider: MultiProtocolProvider,
+    environment: DeployEnvironment,
+    private readonly context: Contexts,
+    private readonly fundingAddresses: FundingAddresses,
+    public readonly logger: Logger,
+    private readonly walletBalanceGauge: Gauge<string>,
+  ) {
+    super(chainName, multiProtocolProvider, fundingAddresses, logger);
+    this.localMultiProvider = multiProtocolProvider.toMultiProvider();
+    this.localMultiProvider.setSigner(chainName, this.getSigner());
+    this.igp = HyperlaneIgp.fromAddressesMap(
+      {
+        [this.chainName]: this.fundingAddresses,
+      },
+      this.localMultiProvider,
+    );
+    this.nativeBridges = {
+      scrollsepolia: {
+        l1ETHGateway: '0x8A54A2347Da2562917304141ab67324615e9866d',
+        l1Messenger: '0x50c7d3e7f7c656493D1D76aaa1a836CedfCBB16A',
+      },
+    };
+    this.L2Chains = L2_CHAINS;
+    this.L2ToL1 = {
+      ...L2_CHAINS.reduce(
+        (acc, chain) => {
+          acc[chain] = L1_CHAIN;
+          return acc;
+        },
+        {} as Record<ChainName, ChainName>,
+      ),
```

Comment:
see other comment about whether l1->l2 bridging is even a desirable thing in the keyfunder